src/storage/src/compacter.rs

```diff
@@ -0,0 +1,691 @@
+            }
+            CompacterMessage::Drop(id) => {
+                if !self.traces.contains_key(&id) {
+                    bail!(
+                        "asked to drop trace for relation {} which doesn't exist.",
+                        id
+                    );
+                }
+
+                let trace = self.traces.remove(&id).expect("trace known to exist");
+                trace.destroy()?;
+            }
+            CompacterMessage::Resume(id, trace) => {
+                if self.traces.contains_key(&id) {
+                    bail!(
+                        "asked to resume trace for relation {} which already exists.",
+                        id
+                    );
+                }
+                self.traces.insert(id, trace);
+            }
+            CompacterMessage::AllowCompaction(id, frontier) => {
+                // We might get a lot of messages for relations we don't
+                // know about here so ignore those.
+                if let Some(trace) = self.traces.get_mut(&id) {
+                    if let Some(compaction_frontier) = trace.compaction {
+                        assert!(frontier >= compaction_frontier);
+                    }
+                    trace.compaction = Some(frontier);
+                }
+            }
+        };
+        Ok(())
+    }
+
+    pub async fn run(&mut self) {
+        let ret = self.compact().await;
+
+        match ret {
+            Ok(_) => (),
+            Err(e) => {
+                error!("Compacter thread encountered an error: {:#}", e);
+                error!("Shutting down compacter thread. No further updates will be persisted.");
+            }
+        }
+    }
+}
+
+/// Read a directory and return all (non-subdirectory) files matching `regex`.
+fn read_dir_regex(path: &Path, regex: &Regex) -> Result<Vec<PathBuf>, anyhow::Error> {
+    let entries = std::fs::read_dir(path).with_context(|| {
+        format!(
+            "failed to read {} looking for {}",
+            path.display(),
+            regex.as_str()
+        )
+    })?;
+    let mut results = vec![];
+    for entry in entries {
+        if let Ok(file) = entry {
+            let path = file.path();
+            let file_name = path.file_name();
+            if file_name.is_none() {
+                continue;
+            }
+
+            let file_name = file_name.unwrap().to_str();
+
+            if file_name.is_none() {
+                continue;
+            }
+
+            let file_name = file_name.unwrap();
+            if regex.is_match(&file_name) {
+                results.push(path.to_path_buf());
+            }
+        }
+    }
+
+    Ok(results)
+}
+
+// Data stored in Batches
+#[derive(Debug, PartialEq)]
+pub enum Message {
+    // (Row, time, diff) tuples
+    Data(Update),
+    // Statements about which timestamps we might still receive data at.
+    Progress(Timestamp),
+}
+
+fn read_message(buf: &[u8], mut offset: usize) -> Option<(Message, usize)> {
```

Comment:
Documenting the signature here might be good. Probably the second returned thing is number of bytes read?

Reply:
What is the `offset`? Or, why shouldn't the caller apply it themselves to `buf`?